central/notifiers/jira/jira.go

```diff
@@ -69,57 +60,172 @@ type jira struct {
-	MaxResults int                      `json:"maxResults"`
-	Total      int                      `json:"total"`
-	IssueTypes []*jiraLib.MetaIssueType `json:"values"`
+	StartAt         int                      `json:"startAt"`
+	MaxResults      int                      `json:"maxResults"`
+	Total           int                      `json:"total"`
+	IssueTypes      []*jiraLib.MetaIssueType `json:"values"`
+	IssueTypesCloud []*jiraLib.MetaIssueType `json:"issueTypes"`
 }
 
-func getIssueTypes(client *jiraLib.Client, project string) ([]*jiraLib.MetaIssueType, error) {
-	// Low level HTTP client call is used here due to the deprecation/removal of the Jira endpoint used by the Jira library
-	// to fetch the CreateMeta data, and there is no API call for the suggested endpoint to use in place of the removed one.
-	// Info here:
-	// https://confluence.atlassian.com/jiracore/createmeta-rest-endpoint-to-be-removed-975040986.html
-	urlPath := fmt.Sprintf("rest/api/2/issue/createmeta/%s/issuetypes", project)
+type issueField struct {
+	Name    string `json:"name"`
+	Key     string `json:"key"`
+	FieldID string `json:"fieldId"`
+}
 
+type issueFieldsResult struct {
+	StartAt          int           `json:"startAt"`
+	MaxResults       int           `json:"maxResults"`
+	Total            int           `json:"total"`
+	IssueFields      []*issueField `json:"values"`
+	IssueFieldsCloud []*issueField `json:"fields"`
+}
+
+type permissionResult struct {
+	Permissions map[string]struct {
+		HavePermission bool
+	}
+}
+
+func callJira(client *jiraLib.Client, urlPath string, result interface{}, startAt int) error {
 	req, err := client.NewRequest("GET", urlPath, nil)
 	if err != nil {
-		return []*jiraLib.MetaIssueType{}, err
+		return errors.Wrap(err, "Failed to create Jira request")
 	}
 
+	values := req.URL.Query()
+	values.Set("startAt", strconv.Itoa(startAt))
+	req.URL.RawQuery = values.Encode()
+
 	resp, err := client.Do(req, nil)
 	if err != nil {
-		return []*jiraLib.MetaIssueType{}, err
+		return errors.Wrap(err, "Failed to successfully make request to Jira")
 	}
 
-	result := &issueTypeResult{}
 	defer utils.IgnoreError(resp.Body.Close)
 	err = json.NewDecoder(resp.Body).Decode(result)
+	if err != nil {
+		return errors.Wrap(err, "Failed to decode JSON response from Jira")
+	}
+
+	return nil
+}
+
+func getIssueTypes(client *jiraLib.Client, project string) ([]*jiraLib.MetaIssueType, error) {
+	urlPath := fmt.Sprintf("rest/api/2/issue/createmeta/%s/issuetypes", project)
+
+	var result issueTypeResult
+
+	err := callJira(client, urlPath, &result, 0)
+
 	if err != nil {
 		return []*jiraLib.MetaIssueType{}, err
 	}
 
-	return result.IssueTypes, nil
+	returnList := make([]*jiraLib.MetaIssueType, 0, result.Total)
+
+	if len(result.IssueTypes) == 0 {
+		returnList = append(returnList, result.IssueTypesCloud...)
+	} else {
+		returnList = append(returnList, result.IssueTypes...)
+	}
+
+	for result.Total != len(returnList) {
```

Comment:
Can you explain what this is doing? I'm a bit confused by it as result is being reset in the loop?

Reply:
`result` is the struct that represents an individual response from the Jira endpoint.  The loop is fetching all the issue types via a sequence of paginated calls.  Each response has a `Total` field as a top level attribute on the JSON response.  This number shouldn't change between calls (unless the set of issue types gets updated in the middle of the sequence of paginated calls).

So once all the items are fetched, the length of `returnList` should equal the `Total` number returned by Jira, indicating to stop making more paginated calls.

Reply:
Okay, can we break if result.Total is < len. Otherwise, reading this there seems like some potential for an infinite loop

Reply:
how about `len(returnList) < result.Total`?